crates/runtime-config/src/lib.rs

```diff
@@ -0,0 +1,127 @@
+            format!(
+                "failed to read runtime config file '{}'",
+                runtime_config_path.display()
+            )
+        })?;
+        let toml = toml::from_str(&file).with_context(|| {
+            format!(
+                "failed to parse runtime config file '{}' as toml",
+                runtime_config_path.display()
+            )
+        })?;
+        let runtime_config: T = TomlRuntimeConfigSource::new(&toml, &key_value_resolver)
+            .try_into()
+            .map_err(Into::into)?;
+
+        Ok(Self {
+            runtime_config,
+            key_value_resolver,
+        })
+    }
+}
+
+/// The TOML based runtime configuration source Spin CLI.
+pub struct TomlRuntimeConfigSource<'a> {
+    table: TomlKeyTracker<'a>,
+    key_value: &'a key_value::RuntimeConfigResolver,
+}
+
+impl<'a> TomlRuntimeConfigSource<'a> {
+    pub fn new(table: &'a toml::Table, key_value: &'a key_value::RuntimeConfigResolver) -> Self {
+        Self {
+            table: TomlKeyTracker::new(table),
+            key_value,
+        }
+    }
+}
+
+impl FactorRuntimeConfigSource<WasiFactor> for TomlRuntimeConfigSource<'_> {
+    fn get_runtime_config(&mut self) -> anyhow::Result<Option<()>> {
+        Ok(None)
+    }
+}
+
+impl FactorRuntimeConfigSource<KeyValueFactor> for TomlRuntimeConfigSource<'_> {
+    fn get_runtime_config(
+        &mut self,
+    ) -> anyhow::Result<Option<spin_factor_key_value::RuntimeConfig>> {
+        self.key_value.resolve_from_toml(Some(self.table.as_ref()))
+    }
+}
+
+impl RuntimeConfigSourceFinalizer for TomlRuntimeConfigSource<'_> {
+    fn finalize(&mut self) -> anyhow::Result<()> {
+        Ok(self.table.validate_all_keys_used()?)
+    }
+}
+
+const DEFAULT_SPIN_STORE_FILENAME: &str = "sqlite_key_value.db";
+
+/// The key-value runtime configuration resolver used by the trigger.
+///
+/// Takes a base path for the local store.
+pub fn key_value_resolver(local_store_base_path: PathBuf) -> key_value::RuntimeConfigResolver {
+    let mut key_value = key_value::RuntimeConfigResolver::new();
+
+    // Register the supported store types.
+    // Unwraps are safe because the store types are known to not overlap.
+    key_value
+        .register_store_type(spin_factor_key_value_spin::SpinKeyValueStore::new(
+            local_store_base_path,
+        ))
+        .unwrap();
+    key_value
+        .register_store_type(spin_factor_key_value_redis::RedisKeyValueStore::new())
+        .unwrap();
+    key_value
+        .register_store_type(spin_factor_key_value_azure::AzureKeyValueStore::new())
+        .unwrap();
+
+    // Add handling of "default" store.
+    key_value.add_default_store(
+        "default",
+        key_value::StoreConfig {
+            type_: spin_factor_key_value_spin::SpinKeyValueStore::RUNTIME_CONFIG_TYPE.to_owned(),
+            config: toml::toml! {
+                path = DEFAULT_SPIN_STORE_FILENAME
+            },
+        },
+    );
+
+    key_value
+}
```

Comment:
Of course it doesn't need to be done in this PR but we'll almost certainly break this out into an e.g. `key_value` module with the next factor added, yeah?

Reply:
Yea most likely.